### PR TITLE
Add ABI sea-ice product to ioda converter

### LINKED ISSUE #1182: More converters for sea-ice obs
- This is mainly for seaice concentration, but we need to look at what else is available (skin temp for ex.)
- Even though there are 8 different instruments/platforms, we should be able to get away with fewer converters, some of the files structure seem exactly the same across platform/inst.
 
Copy/paste of @ilianagenkova 's README:
```
1.ABI-L2-AICEF
2.AMSR2_SEAICE_NH
3.AMSR2_SEAICE_SH  
4.JPSSRR_IceConcentration
5.MIRS_AMSU-MHS_IMG
6.MIRS_ATMS_IMG  
7.MIRS_GMI_IMG
8.MIRS_High_Resolution_IMG 

The sub names (directory) are as follows
1.EMC_ABI-L2-AICEF (PULL/ABI_L2_AICEF)
2.EMC_AMSR2_SEAICE_NH (NDE/AMSR2)
3.EMC_AMSR2_SEAICE_SH (NDE/AMSR2)
4.EMC_JPSSRR_IceConcentration (PULL/JPSSRR_IceConcentration)
5.EMC_MIRS_AMSU-MHS_IMG (PULL/MIRS_AMSU)

8.EMC_MIRS_High_Resolution_IMG (PULL/MIRS_High_Resolution_IMG)
7.EMC_MIRS_GMI_IMG (PULL/MIRS_GMI_IMG)
6a.EMC_MIRS_ATMS_IMG_J1 (PULL/MIRS_ATMS_IMG/J1)
6b.EMC_MIRS_ATMS_IMG_J2 (PULL/MIRS_ATMS_IMG/J2)
6c.EMC_MIRS_ATMS_IMG_SNPP (PULL/MIRS_ATMS_IMG/SNPP)

For (6) MIRS_ATMS_IMG, 
3 subs were created for each satellite because each satellite produces ~2600 files/day.
If files from all 3 satellites were in the same directory, retrieval would have been slow.
```

TODO:
- [x] document the list of retrieved parameters (ice concentration, ice temp, ...)
- [x] make sure all the intrument/platform provided by @ilianagenkova have at least seaice concentration as retrieved obs. Communicate to Iliana all the instruments/platforms that we want as part of the obs processing.
- [x] design the converter(s)
   - [x] MIRS Instruments (NPP)
   - [x] ABI Instruments
   - [x] JPSSRR Instruments
   
- [x] test within a cycling experiment with S-NPP
- [ ] test within a cycling experiment with ABI
- [x] test within a cycling experiment with JPSSRR

- [x] Final Test with all converters

### PR DESCRIPTION
**Pull request by apchoiCMD**
#### Description
- Task for adding a new ioda converter for new `ABI` type sea-ice concentration
- For sea-ice obs, AMSR2, MIRS, JPSSRR and ABI products will be used
- For the part of conversion of metadata, the reference is [here](https://www.goes-r.gov/products/docs/PUG-L2+-vol5.pdf)

Close #1182

### CODE REVIEW

**Review by AndrewEichmann-NOAA — DISMISSED** [review]
GLTM

File: utils/test/testinput/gdas_icecabi2ioda.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+provider: ABI 
+window begin: 2024-06-18T15:00:00Z
+window end: 2024-06-19T21:00:00Z
+output file: icec_abi.ioda.nc
+#ocean basin: RECCAP2_region_masks_all_v20221025.nc
```
Comment (JohnSteffen-NOAA) [inline utils/test/testinput/gdas_icecabi2ioda.yaml:5]: Is this supposed to be commented out? If so, is it worth deleting or do we want to keep it for future testing?

**Review by JohnSteffen-NOAA — DISMISSED** [review]
Looks good. Thanks for the walk through. Also, you may need to merge the latest develop branch first.

File: utils/test/testinput/gdas_icecabi2ioda.yaml (hunk 1 above)
Comment (apchoiCMD) [reply utils/test/testinput/gdas_icecabi2ioda.yaml:5]: It is currently not part of the ctest, and needs to be used when processing obs.

File: utils/obsproc/util.h [hunk 2]
```diff
@@ -254,5 +256,58 @@ namespace gdasapp {
         }
       };
     }  // namespace iodavars
+
+    namespace utils {
+
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h:261]: Add a TODO to move this as a private method to the icecabi2ioda class.

File: utils/test/prepdata.sh [hunk 3]
```diff
@@ -2,6 +2,17 @@
 # called for test_gdasapp_util_prepdata, and by
 # test/soca/gw/setup_obsproc.sh for test_gdasapp_soca_setup_obsproc
 
+# Get the machine hostname
```
Comment (guillaumevernieres) [inline utils/test/prepdata.sh:6]: add a TODO to point to the experimental obs instead of cdl to nc

File: utils/obsproc/IcecAbi2Ioda.h [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+      // Get Ice_Concentration obs values
+      std::vector<uint16_t> IcecObsVal(iodaVars.location_);
+      ncFile.getVar("IceConc").getVar(IcecObsVal.data());
+      float IcecOffSet;
+      ncFile.getVar("IceConc").getAtt("add_offset").getValues(&IcecOffSet);
+      float IcecScaleFactor;
+      ncFile.getVar("IceConc").getAtt("scale_factor").getValues(&IcecScaleFactor);
+
+      // Get Ice_Temp obs values
```
Comment (apchoiCMD) [inline utils/obsproc/IcecAbi2Ioda.h:143]: Add the comments that is not used for now

**Comment by apchoiCMD** [comment]
Updated results with original file;

combiled_crop

**Comment by apchoiCMD** [comment]
Tested ABI ioda converter within cycling test:

ABI_cycling_plot

File: utils/obsproc/IcecAbi2Ioda.h [hunk 5]
```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include <complex>
+#include <ctime>
+#include <iomanip>
+#include <iostream>
+#include <map>
+#include <netcdf>    // NOLINT (using C API)
+#include <sstream>
```
Comment (guillaumevernieres) [inline utils/obsproc/IcecAbi2Ioda.h]: is `sstream` used?

File: utils/obsproc/util.h [hunk 6]
```diff
@@ -254,5 +255,59 @@ namespace gdasapp {
...
     }  // namespace iodavars
+
+    // TODO(Mindo): To move below as a private method to the iceabi2ioda class
+    namespace utils {
+
+       // Calculate latitude and longitude from GOES ABI fixed grid projection data
+       // GOES ABI fixed grid projection is a map projection relative to the GOES satellite
+       // Units: latitude in °N (°S < 0), longitude in °E (°W < 0)
+       // See GOES-R Product User Guide (PUG) Volume 5 (L2 products) Section 4.2.8 (p58)
+       void calculate_degrees(
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: up to you and I won't block your PR for this, but it's a weird name for a function and I think we've been camel casing the method's names too.
`degrees` is vague and has multiple meanings, something like `calcGeoLoc` or `abiToGeoLoc` ... ?

File: utils/obsproc/util.h [hunk 7]
```diff
@@ -254,5 +255,59 @@ namespace gdasapp {
...
+           double lambda_0 = (lon_origin * M_PI) / 180.0;
+
+           abi_lat.resize(sizeY, std::vector<double>(sizeX));
+           abi_lon.resize(sizeY, std::vector<double>(sizeX));
+
+           for (int i = 0; i < sizeY; ++i) {
+             for (int j = 0; j < sizeX; ++j) {
+                double x = x_coordinate_2d[i][j];
+                double y = y_coordinate_2d[i][j];
+                double a_var = std::pow(std::sin(x), 2.0)
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: I have no clue how much time is spent on these loops, but you could cache `cos(x or y)` and `sin(x or y)`, it looks like it's used in multiple places, you might save a micro second!

File: utils/obsproc/util.h [hunk 8]
```diff
@@ -254,5 +255,59 @@ namespace gdasapp {
...
+                              + std::pow(std::cos(x), 2.0) * (std::pow(std::cos(y), 2.0)
+                              + ((r_eq * r_eq) / (r_pol * r_pol)) * std::pow(std::sin(y), 2.0));
+                double b_var = -2.0 * H * std::cos(x) * std::cos(y);
+                double c_var = (H * H) - (r_eq * r_eq);
+                double discriminant = (b_var * b_var) - (4.0 * a_var * c_var);
+                double r_s = (-b_var - std::sqrt(discriminant)) / (2.0 * a_var);
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: check that `discriminant` is >0 before computing the `sqrt`

File: utils/obsproc/util.h [hunk 9]
```diff
@@ -254,5 +255,59 @@ namespace gdasapp {
...
+                double s_x = r_s * std::cos(x) * std::cos(y);
+                double s_y = -r_s * std::sin(x);
+                double s_z = r_s * std::cos(x) * std::sin(y);
+
+                abi_lat[i][j] = (180.0 / M_PI) * (std::atan(((r_eq * r_eq) / (r_pol * r_pol))
+                               * (s_z / std::sqrt(((H - s_x) * (H - s_x)) + (s_y * s_y)))));
+                abi_lon[i][j] = (lambda_0 - std::atan(s_y / (H - s_x))) * (180.0 / M_PI);
+
+                // Handle invalid values
+                if (discriminant < 0 || std::isnan(abi_lat[i][j]) || std::isnan(abi_lon[i][j])) {
```
Comment (guillaumevernieres) [inline utils/obsproc/util.h]: see comment above

File: utils/obsproc/IcecAbi2Ioda.h [hunk 10]
```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include <complex>
```
Comment (guillaumevernieres) [inline utils/obsproc/IcecAbi2Ioda.h]: I don't think `complex` is used ...

File: utils/obsproc/IcecAbi2Ioda.h [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+#include <netcdf>    // NOLINT (using C API)
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "eckit/config/LocalConfiguration.h"
+
+#include <Eigen/Dense>    // NOLINT
+
+#include "ioda/../../../../core/IodaUtils.h"
```
Comment (guillaumevernieres) [inline utils/obsproc/IcecAbi2Ioda.h]: Not related to this PR, but we need to fix this at some point. Can you add a TODO here?

**Review by guillaumevernieres — COMMENTED** [review]
Thanks @apchoiCMD , see comments.

File: utils/obsproc/util.h (hunk 6 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: I would love to use `abiToGeoLoc`-

File: utils/obsproc/IcecAbi2Ioda.h (hunk 5 above)
Comment (apchoiCMD) [reply utils/obsproc/IcecAbi2Ioda.h]: Good catch!

File: utils/obsproc/IcecAbi2Ioda.h (hunk 10 above)
Comment (apchoiCMD) [reply utils/obsproc/IcecAbi2Ioda.h]: Yep, thanks!

File: utils/obsproc/IcecAbi2Ioda.h (hunk 11 above)
Comment (apchoiCMD) [reply utils/obsproc/IcecAbi2Ioda.h]: Added

File: utils/obsproc/util.h (hunk 8 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: Added

File: utils/obsproc/util.h (hunk 7 above)
Comment (apchoiCMD) [reply utils/obsproc/util.h]: Hope this saves a micro second-

**Review by JohnSteffen-NOAA — APPROVED** [review]

**Comment by guillaumevernieres** [comment]
tested by @JohnSteffen-NOAA , does what it's supposed to.

**Pull request merged**